lightning/src/ln/router.rs

```diff
@@ -1864,4 +1868,672 @@ mod tests {
+		};
+		unsigned_announcement.chain_hash = genesis_block(Network::Testnet).header.bitcoin_hash();
+
+		unsigned_announcement.node_id_1 = PublicKey::from_secret_key(&secp_ctx, node_2_privkey);
+		msghash = hash_to_message!(&Sha256dHash::hash(&unsigned_announcement.encode()[..])[..]);
+		let channel_to_itself_announcement = ChannelAnnouncement {
+			node_signature_1: secp_ctx.sign(&msghash, node_1_privkey),
+			node_signature_2: secp_ctx.sign(&msghash, node_1_privkey),
+			bitcoin_signature_1: secp_ctx.sign(&msghash, node_1_btckey),
+			bitcoin_signature_2: secp_ctx.sign(&msghash, node_2_btckey),
+			contents: unsigned_announcement.clone(),
+		};
+	 	match router.handle_channel_announcement(&channel_to_itself_announcement) {
+			Ok(_) => panic!(),
+			Err(e) => assert_eq!(e.err, "Channel announcement node had a channel with itself")
+		};
+	}
+
+	#[test]
+	fn handling_channel_update() {
+		let (secp_ctx, _, router) = create_router();
+		let secret_key = &SecretKey::from_slice(&hex::decode("0202020202020202020202020202020202020202020202020202020202020202").unwrap()[..]).unwrap();
+		let public_key = PublicKey::from_secret_key(&secp_ctx, secret_key);
+
+		let zero_hash = Sha256dHash::hash(&[0; 32]);
+		let channel_id = 0;
+		let channel_time = 100;
+		let channel_key = NetworkMap::get_key(channel_id, genesis_block(Network::Testnet).header.bitcoin_hash());
+
+		{
+			let mut network = router.network_map.write().unwrap();
+			network.channels.insert(channel_key, ChannelInfo {
+				features: ChannelFeatures::empty(),
+				one_to_two: DirectionalChannelInfo {
+					src_node_id: public_key.clone(),
+					last_update: channel_time,
+					enabled: false,
+					cltv_expiry_delta: u16::max_value(),
+					htlc_minimum_msat: 0,
+					fee_base_msat: u32::max_value(),
+					fee_proportional_millionths: u32::max_value(),
+					last_update_message: None,
+				}, two_to_one: DirectionalChannelInfo {
+					src_node_id: public_key.clone(),
+					last_update: channel_time,
+					enabled: true,
+					cltv_expiry_delta: 0,
+					htlc_minimum_msat: 0,
+					fee_base_msat: u32::max_value(),
+					fee_proportional_millionths: 0,
+					last_update_message: None,
+				},
+				announcement_message: None,
+			});
+		}
+
+		let mut unsigned_channel_update = UnsignedChannelUpdate {
+			chain_hash: genesis_block(Network::Testnet).header.bitcoin_hash(),
+			short_channel_id: channel_id,
+			timestamp: channel_time + 100,
+			flags: 0,
+			cltv_expiry_delta: 144,
+			htlc_minimum_msat: 1000000,
+			fee_base_msat: 10000,
+			fee_proportional_millionths: 20,
+			excess_data: Vec::new()
+		};
+		let mut msghash = hash_to_message!(&Sha256dHash::hash(&unsigned_channel_update.encode()[..])[..]);
+		let valid_channel_update = ChannelUpdate {
+			signature: secp_ctx.sign(&msghash, secret_key),
+			contents: unsigned_channel_update.clone()
+		};
+
+		// TODO: local node (not test) causes panic because remote node does not exist.
```

Comment:
This state should never be reachable without modifying internal state :p. If you updated the above to instead insert the channels by announcing a message then this wouldn't panic :).